helix-term/src/ui/editor.rs

```diff
@@ -1362,17 +1362,11 @@ impl Component for EditorView {
                 // if the focused view still exists and wasn't closed
                 if cx.editor.tree.contains(focus) {
                     let config = cx.editor.config();
-                    let mode = cx.editor.mode();
                     let view = view_mut!(cx.editor, focus);
                     let doc = doc_mut!(cx.editor, &view.doc);
 
                     view.ensure_cursor_in_view(doc, config.scrolloff);
-
-                    // Store a history state if not in insert mode. This also takes care of
-                    // committing changes when leaving insert mode.
-                    if mode != Mode::Insert {
-                        doc.append_changes_to_history(view);
-                    }
```

Comment:
This insert-mode check is important: without gating this on `mode != Mode::Insert` we get an undo checkpoint for every modification in insert mode. For example `ifoo<esc>` has an undo checkpoint for each character in `foo`.